This is the second session on basic usage of Python 3, which is focussed on general usage.

## Course Structure

### Session 2
**Loops in Python**
* Conditional statements: if, else
* Loop exits: Pass,break, continue
* List comprehension


**Reading/writing to/from a file**
* Line by line
* Using Pandas

**Hands-on Pandas**
* Case Study

**Homework**


## For loop
<font color=blue>**Unlike many languages, Python's loops do not intrinsically iterate over integers, but rather elements in sequences, like lists and tuples.<br>
This adds efficiency**</font> <br>

<font color=blue>**Syntax:**</font> <br>

for element in sequence:
> commands

In [1]:
for i in [3, "hello", 9.5]:
    print(i)

3
hello
9.5


<font color=blue>Notice that anything falling within the loop is indented beneath the first line, similar to if statements</font> <br>

In [2]:
for i in (2.3, [8, 9, 10], {"city":"Santa Barbara"}):
    print(i)

2.3
[8, 9, 10]
{'city': 'Santa Barbara'}


In [6]:
for i in ([2.3], [8, 9, 10], {"city","Santa Barbara"}):
    for j in i:
        print(j)

2.3
8
9
10
Santa Barbara
city


<font color=blue>The items in the iterable do not need to be the same type.<br>
In each case, the variable 'i' is given the value of the current list element, and the loop proceeds over these in sequence.<br> 
Also, one does not have to use the variable i; any variable name will do, but if an existing variable is used, its value will be overwritten by the loop.</font> <br>

<font color=blue>**It is very easy to loop over a part of a list using slicing:**</font> <br>

In [9]:
l = [4, 6, 7, 8, 10]

for i in l[2:]:
    i = i*5
    print(i)

35
40
50


<font color=blue>**Using Python's multiple assignment capabilities, it is possible to iterate over more than one value at a time:**</font> <br>

In [11]:
jack = [(1, 2), (3, 4), (5, 6)]

for (i,j) in jack:
    sum_ = i+j
    print(sum_)

3
7
11


In [17]:
jill = [('indiana', 'jones'), ('Mr', 'India'), ('Cast', 'Away')]
    
for (i,j) in jill:
    sum_ = i+j
    print(sum_)

indianajones
MrIndia
CastAway


<font color=blue>**It is possible to iterate over sequences of numbers using the range function: <br>
 Syntax is:
    **</font> <br>


In [ ]:
for i in range(start, end, step):
    do something using i

In [50]:
for i in range(0,23,2):
    print(i)

0
2
4
6
8
10
12
14
16
18
20
22


In [49]:
for i in range(20,0,-2):
    print(i)

20
18
16
14
12
10
8
6
4
2


In [16]:
l = [8, 10, 12]

for (i, val) in enumerate(l):
    print("The %ith element in the list is %d" % (i, val))

The 0th element in the list is 8
The 1th element in the list is 10
The 2th element in the list is 12


<font color=blue>**It is also possible to iterate over two lists simultaneously using the zip function:**</font> <br>

In [23]:
l1 = [1, 2, 3]
l2 = [0, 6, 8]

for (a) in zip(l1, l2):
    print(a)

(1, 0)
(2, 6)
(3, 8)


<font color=blue>**It simply takes two or more lists and groups them together, making tuples of corresponding list elements**</font> <br>

<font color=blue>**For loop inside the list:** <br> Python provides a convenient syntax for creating new lists from existing lists, tuples. These list comprehensions have the general form:</font> <br>

[expression for object in iterable]

In [51]:
[i*i for i in range(5)]

[0, 1, 4, 9, 16]

<font color=blue>In the expression above, elements from the list created by the range function are accessed in sequence and assigned to the variable i. The new list then takes each element and squares it.</font> <br>

In [53]:
[q**(0.5) for q in (4,9,16)]

[2.0, 3.0, 4.0]

In [54]:
[character for character in "Python"]

['P', 'y', 't', 'h', 'o', 'n']

In [55]:
[k % 2 == 0 for k in range(5)]

[True, False, True, False, True]

<font color=blue>**It is also possible to filter items in list comprehensions using if statements:**</font> <br>

In [57]:
[s for s in ["blue", "red", "green", "yellow"] if "l" in s]

['blue', 'yellow']

In [58]:
[s[0] for s in ["blue", "red", "green", "yellow"] if "l" in s]

['b', 'y']

### logical statements

<font color=blue>**While loops repeat as long as a certain boolean condition is met:**</font> <br>


In [61]:
count = 0
while count < 5:
    print(count)
    count += 1  # This is the same as count = count + 1

0
1
2
3
4


<font color=blue>'break' is used to exit a for loop or a while loop, whereas 'continue' is used to skip the current block, and return to the "for" or "while" statement.</font> <br>

In [81]:
number = 0

for number in range(10):
    if number == 5:
        break    # break here

    print('Number is ' + str(number))

Number is 0
Number is 1
Number is 2
Number is 3
Number is 4
Out of loop


<font color=blue>the variable number is initialized at 0. Then a for statement constructs the loop as long as the variable 'number' is less than 10. Within the for loop, there is an if statement that presents the condition that if the variable number is equivalent to the integer 5, then the loop will break</font> <br>

In [82]:
number = 0

for number in range(10):
    if number == 5:
        continue    # continue here

    print('Number is ' + str(number))

Number is 0
Number is 1
Number is 2
Number is 3
Number is 4
Number is 6
Number is 7
Number is 8
Number is 9


<font color=blue>continue forces the loop to start at the next iteration while pass means "there is no code to execute here" and will continue through the remainder or the loop body</font> <br>

In [83]:
number = 0

for number in range(10):
    if number == 5:
        pass    # pass here

    print('Number is ' + str(number))

Number is 0
Number is 1
Number is 2
Number is 3
Number is 4
Number is 5
Number is 6
Number is 7
Number is 8
Number is 9


</font> <br> **so actually where 'pass' is used**</font> <br>

In [94]:
x = 10

for number in range(5,-5,-1):
    division = x/number

    print('Number is ' + str(number))
    

Number is 5
Number is 4
Number is 3
Number is 2
Number is 1


ZeroDivisionError: division by zero

In [96]:
x = 10
for number in range(5,-5,-1):
    try:
        division = x/number
        print('Number is ' + str(number))
    except:
        print('division by zero')
        pass
    

Number is 5
Number is 4
Number is 3
Number is 2
Number is 1
division by zero
Number is -1
Number is -2
Number is -3
Number is -4


### Python Classes

<font color=blue>**Python, allows us to create new object types called classes. <br> 
We can then use these classes to create new objects of our own design. In the following example, we create a new class that describes a Car:**</font> <br>

In [38]:
class Car:
    def __init__(self, Distance, Time = 12.0, Name = 'Maruti'):
        self.Name = Name 
        self.Distance = Distance 
        self.Time = Time 
    def Average(self):
        avg = self.Distance / self.Time
        return avg

In [39]:
a = Car(Name = 'Mercedes', Distance = 30, Time = 2.0)

In [40]:
b = Car(10)

In [42]:
a.Name, a.Average()

('Mercedes', 15.0)

In [43]:
b.Name

'Maruti'

In [44]:
b.Average()

0.8333333333333334

<font color=blue>In this example, the class statement indicates the creation of a new class called Car; <br>
All definitions for this class must be indented underneath it.
The first definition is for a special function called __init__ that is a constructor for the class, meaning this function is automatically executed by Python every time a new object of type AtomClass is created.<br> There are actually many special functions that can be defined for a class; each of these begins and ends with two underscore marks.<br>
Notice that the first argument to the __init__ function is the object self<font color=blue>

## Case Study

### Read file using Python- line by line

In [140]:
file_name = "C:/Users/singhn25/Downloads/Python_Sessions/au-towns-sample.csv"
file = open(file_name)
result = file.read().split("\n")

In [141]:
result

['id,town,urban_area,state_code,state,postcode,type,latitude,longitude,elevation,population,median_income,area_sq_km,local_government_area,region,time_zone',
 '1,Aarons Pass,,NSW,New South Wales,2850,Rural locality,-32.86328,149.80375,804,22,29900,82.764,Mid-Western Regional (Area),Central West,Australia/Sydney',
 '2,Abba River,,WA,Western Australia,6280,Rural locality,-33.68488,115.46334,23,68,35100,28.516,Busselton (City),Bunbury,Australia/Perth',
 '3,Abbey,Busselton,WA,Western Australia,6280,Suburb,-33.66077,115.25863,7,1275,34580,2.644,Busselton (City),Bunbury,Australia/Perth',
 '4,Abbeyard,,VIC,Victoria,3737,Rural locality,-36.96978,146.75244,415,0,0,327.501,Alpine (Shire),Hume,Australia/Melbourne',
 '5,Abbeywood,,QLD,Queensland,4613,Rural locality,-26.10688,151.6288,429,23,16484,35.035,South Burnett (Regional Council),Wide Bay,Australia/Brisbane',
 '6,Abbotsbury,Sydney,NSW,New South Wales,2176,Suburb,-33.87754,150.86777,80,4253,37492,4.979,Fairfield (City),Sydney - South West,Aus

### Introduction: Pandas

<font color=blue> pandas is a fast, powerful, flexible and easy to use open source data analysis and manipulation tool, built on top of the Python programming language <font color=blue>

In [142]:
import pandas as pd

<font color=blue> read file: <font color=blue>

In [143]:
df = pd.read_csv(file_name)

<font color=blue> see first, say 4 values <font color=blue>

In [144]:
df.head(4)

id         town urban_area state_code              state  postcode  \
0   1  Aarons Pass        NaN        NSW    New South Wales      2850   
1   2   Abba River        NaN         WA  Western Australia      6280   
2   3        Abbey  Busselton         WA  Western Australia      6280   
3   4     Abbeyard        NaN        VIC           Victoria      3737   

             type  latitude  longitude  elevation  population  median_income  \
0  Rural locality -32.86328  149.80375        804          22          29900   
1  Rural locality -33.68488  115.46334         23          68          35100   
2          Suburb -33.66077  115.25863          7        1275          34580   
3  Rural locality -36.96978  146.75244        415           0              0   

   area_sq_km        local_government_area        region            time_zone  
0      82.764  Mid-Western Regional (Area)  Central West     Australia/Sydney  
1      28.516             Busselton (City)       Bunbury      Australia/Perth  
2       2.644             Busselton (City)       Bunbury      Australia/Perth  
3     327.501               Alpine (Shire)          Hume  Australia/Melbourne

In [145]:
df.tail(4)

id       town          urban_area state_code            state  postcode  \
458  459    Axedale                 NaN        VIC         Victoria      3551   
459  460  Aylmerton  Bowral - Mittagong        NSW  New South Wales      2575   
460  461        Ayr                 NaN        QLD       Queensland      4807   
461  462    Ayrford                 NaN        VIC         Victoria      3268   

               type  latitude  longitude  elevation  population  \
458  Urban locality -36.78792  144.50413        174         802   
459          Suburb -34.41907  150.49274        599         194   
460  Urban locality -19.57605  147.40517         16        8738   
461  Rural locality -38.41566  142.85945         97          44   

     median_income  area_sq_km   local_government_area  \
458          34268     113.614  Greater Bendigo (City)   
459          34008       2.796    Wingecarribee (Area)   
460          31356      28.473        Burdekin (Shire)   
461          44876      12.900           Moyne (Shire)   

                                region            time_zone  
458                            Bendigo  Australia/Melbourne  
459  Southern Highlands and Shoalhaven     Australia/Sydney  
460                         Townsville   Australia/Brisbane  
461         Warrnambool and South West  Australia/Melbourne

<font color=blue> access a column <font color=blue>

In [152]:
df['town']

0      Aarons Pass
1       Abba River
2            Abbey
3         Abbeyard
4        Abbeywood
          ...     
457      Axe Creek
458        Axedale
459      Aylmerton
460            Ayr
461        Ayrford
Name: town, Length: 462, dtype: object

<font color=blue> **now let's perform some of the operation again that we have already done today:** <font color=blue>

<font color=blue> to add a string xyz to each name  <font color=blue>

In [148]:
['xyz -'+n for n in df['town']]

['xyz -Aarons Pass',
 'xyz -Abba River',
 'xyz -Abbey',
 'xyz -Abbeyard',
 'xyz -Abbeywood',
 'xyz -Abbotsbury',
 'xyz -Abbotsford',
 'xyz -Abbotsford',
 'xyz -Abbotsford',
 'xyz -Abbotsham',
 'xyz -Abels Bay',
 'xyz -Abercorn',
 'xyz -Abercrombie',
 'xyz -Abercrombie River',
 'xyz -Aberdare',
 'xyz -Aberdeen',
 'xyz -Aberdeen',
 'xyz -Aberfeldie',
 'xyz -Aberfeldy',
 'xyz -Aberfoyle',
 'xyz -Aberfoyle Park',
 'xyz -Aberglasslyn',
 'xyz -Abergowrie',
 'xyz -Abermain',
 'xyz -Abernethy',
 'xyz -Abingdon Downs',
 'xyz -Abington',
 'xyz -Abington',
 'xyz -Acacia Creek',
 'xyz -Acacia Gardens',
 'xyz -Acacia Hills',
 'xyz -Acacia Hills',
 'xyz -Acacia Ridge',
 'xyz -Acheron',
 'xyz -Acland',
 'xyz -Acton',
 'xyz -Acton',
 'xyz -Acton Park',
 'xyz -Acton Park',
 'xyz -Ada',
 'xyz -Adaminaby',
 'xyz -Adams Estate',
 'xyz -Adamstown',
 'xyz -Adamstown Heights',
 'xyz -Adamsvale',
 'xyz -Adare',
 'xyz -Adavale',
 'xyz -Addington',
 'xyz -Adelaide',
 'xyz -Adelaide Lead',
 'xyz -Adelaide Park',

<font color=blue> to find names that ends with 'ss' <font color=blue>

In [149]:
[n for n in df['town'] if n[-2:] == 'ss']

['Aarons Pass', 'Applecross', 'Ardross']

<font color=blue> to join name with state <font color=blue>

In [129]:
[x for x in zip(df['name'], df['state'])]

[('Aarons Pass', 'New South Wales'),
 ('Abba River', 'Western Australia'),
 ('Abbey', 'Western Australia'),
 ('Abbeyard', 'Victoria'),
 ('Abbeywood', 'Queensland'),
 ('Abbotsbury', 'New South Wales'),
 ('Abbotsford', 'New South Wales'),
 ('Abbotsford', 'Queensland'),
 ('Abbotsford', 'Victoria'),
 ('Abbotsham', 'Tasmania'),
 ('Abels Bay', 'Tasmania'),
 ('Abercorn', 'Queensland'),
 ('Abercrombie', 'New South Wales'),
 ('Abercrombie River', 'New South Wales'),
 ('Aberdare', 'New South Wales'),
 ('Aberdeen', 'New South Wales'),
 ('Aberdeen', 'Tasmania'),
 ('Aberfeldie', 'Victoria'),
 ('Aberfeldy', 'Victoria'),
 ('Aberfoyle', 'New South Wales'),
 ('Aberfoyle Park', 'South Australia'),
 ('Aberglasslyn', 'New South Wales'),
 ('Abergowrie', 'Queensland'),
 ('Abermain', 'New South Wales'),
 ('Abernethy', 'New South Wales'),
 ('Abingdon Downs', 'Queensland'),
 ('Abington', 'New South Wales'),
 ('Abington', 'Queensland'),
 ('Acacia Creek', 'New South Wales'),
 ('Acacia Gardens', 'New South Wales'

<font color=blue> pick all the names and pupulation having state code= 3737 <font color=blue>

In [150]:
[(x,y) for x,y in zip(df['town'], df['population']) if y > 1000]

[('Abbey', 1275),
 ('Abbotsbury', 4253),
 ('Abbotsford', 5373),
 ('Abbotsford', 8184),
 ('Abercrombie', 1109),
 ('Aberdare', 2473),
 ('Aberdeen', 2084),
 ('Aberfeldie', 3897),
 ('Aberfoyle Park', 11039),
 ('Aberglasslyn', 5401),
 ('Abermain', 2337),
 ('Acacia Gardens', 3798),
 ('Acacia Ridge', 7429),
 ('Acton', 2086),
 ('Acton', 1349),
 ('Acton Park', 2078),
 ('Adamstown', 6044),
 ('Adamstown Heights', 5299),
 ('Adelaide', 15115),
 ('Agnes Water', 2210),
 ('Ainslie', 5189),
 ('Airds', 2926),
 ('Airlie Beach', 1208),
 ('Aitkenvale', 4790),
 ('Alawa', 2127),
 ('Albanvale', 5491),
 ('Albany', 1313),
 ('Albany Creek', 15769),
 ('Alberton', 1818),
 ('Albert Park', 1724),
 ('Albert Park', 6215),
 ('Albion', 2296),
 ('Albion', 4731),
 ('Albion Park', 13316),
 ('Albion Park Rail', 6918),
 ('Albury', 4804),
 ('Aldavilla', 1230),
 ('Alderley', 6117),
 ('Aldershot', 1152),
 ('Aldgate', 3361),
 ('Aldinga Beach', 10557),
 ('Alexander Heights', 7924),
 ('Alexandra', 2695),
 ('Alexandra Headland', 39

In [151]:
[(x,y) for x,y in zip(df['town'], df['postcode']) if y == 3268]

[('Ayrford', 3268)]

In [185]:
[w.split('(')[1] for w in df['local_government_area']]

IndexError: list index out of range

### let's do more with Pandas in next session and see how it is used @ CBA

## Home work

1. Write a program to remove vowels word from list: ['Python', 'India', 'Australia',crypt']
- using loops (continue)
- using list comprehension
2. refer to the attached dataset and answer following:
- Find number of town whose state's spelling ends with 'land'
- Find sum of the population of states having median income less than 30,000
- Find total population of each state 